util/dvsim/Testplan.py

```diff
@@ -0,0 +1,686 @@
+                f"  Description:\n{desc}\n")
+
+    def _validate(self):
+        """Runs some basic consistency checks."""
+        if not self.name:
+            raise ValueError(f"Error: {self.kind.capitalize()} name cannot "
+                             f"be empty:\n{self}")
+
+
+class Covergroup(TestplanElement):
+    """A coverage model item.
+
+    The list of covergroups defines the coverage model for the design. Each
+    entry captures the name of the covergroup (suffixed with _cg) and a brief
+    description describing what functionality is covered. It is recommended to
+    include individual coverpoints and crosses in the description.
+    """
+    kind = "covergroup"
+
+
+class Testpoint(TestplanElement):
+    """An testcase entry in the testplan.
+
+    A testpoint maps to a unique design feature that is planned to be verified.
+    It captures following information:
+    - name of the planned test
+    - a brief description indicating intent, stimulus and checking procedure
+    - the targeted milestone
+    - the list of actual developed tests that verifies it
+    """
+    kind = "testpoint"
+    fields = TestplanElement.fields + ("milestone", "tests")
+
+    # Verification milestones.
+    milestones = ("N.A.", "V1", "V2", "V3")
+
+    def __init__(self, raw_dict):
+        super().__init__(raw_dict)
+
+        # List of TestResult objects indicating test results mapped to this
+        # testpoint.
+        self.test_results = []
+
+    def __str__(self):
+        return super().__str__() + (f"  Milestone: {self.milestone}\n"
+                                    f"  Tests: {self.tests}\n")
+
+    def _validate(self):
+        super()._validate()
+        if self.milestone not in Testpoint.milestones:
+            raise ValueError(f"Testpoint milestone {self.milestone} is "
+                             f"invalid:\n{self}\nLegal values: "
+                             f"Testpoint.milestones")
+
+    def do_substitutions(self, substitutions):
+        '''Substitute {wildcards} in tests
+
+        If tests have {wildcards}, they are substituted with the 'correct'
+        values using the key=value pairs provided by the substitutions arg. If
+        wildcards are present but no replacement is available, then the
+        wildcards are replaced with an empty string.
+
+        substitutions is a dict of key-value pairs.
+        '''
+        resolved_tests = []
+        for test in self.tests:
+            match = re.findall(r"{([A-Za-z0-9\_]+)}", test)
+            if not match:
+                resolved_tests.append(test)
+                continue
+
+            # 'match' is a list of wildcards used in the test. Get their
+            # corresponding values.
+            subst = {item: substitutions.get(item, "") for item in match}
+
+            resolved = [test]
+            for item, value in subst.items():
+                if isinstance(value, list):
+                    for v in value:
```

Comment:
May make this a list comprehension as well? I don't recall if the style guide frowns upon nested comprehensions.